### PR TITLE
Roles search by keyword

### LINKED ISSUE #7362: api call listRoles name filter not working as expected
##### ISSUE TYPE

 * Bug Report

##### COMPONENT NAME

~~~
API
~~~

##### CLOUDSTACK VERSION

~~~
4.18.0.0
~~~

##### SUMMARY

When using parameter name with api call listRoles it lists all roles which have the value of name in the actual name instead of the roles with the exact value of name. Parameter keyword should be used to filter when wanting to find all roles where the value is part of the name.

##### STEPS TO REPRODUCE

~~~
Create a new role with name "New Domain Admin".
Do the following api call: /client/api/?command=listRoles&name=Domain%20Admin&response=json
~~~

##### EXPECTED RESULTS

~~~
You should only get 1 result back. The result should be the default "Domain Admin" role.
~~~

##### ACTUAL RESULTS

~~~
You get 2 results back. The default "Domain Admin" and the new created "New Domain Admin".
~~~

### PR DESCRIPTION
**Pull request by harikrishna-patnala**
### Description

This PR fixes #7362 and also other search criteria to use the name as an exact search where keyword is also there.

Made UI changes for roles search to make use of keyword instead of name.

### Types of changes

- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] New feature (non-breaking change which adds functionality)
- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] Enhancement (improves an existing feature and functionality)
- [x] Cleanup (Code refactoring and cleanup, that may add test cases)

#### Feature/Enhancement Scale

- [ ] Major
- [ ] Minor

#### Bug Severity

- [ ] BLOCKER
- [ ] Critical
- [ ] Major
- [ ] Minor
- [ ] Trivial

### CODE REVIEW

**Comment by harikrishna-patnala** [comment]
@blueorangutan package

**Comment by harikrishna-patnala** [comment]
@blueorangutan package

**Comment by harikrishna-patnala** [comment]
@blueorangutan package

**Comment by harikrishna-patnala** [comment]
@blueorangutan test

File: plugins/metrics/src/main/java/org/apache/cloudstack/metrics/MetricsServiceImpl.java [hunk 1]
```diff
@@ -340,7 +340,7 @@ protected Pair<List<VolumeVO>, Integer> searchForVolumesInternal(ListVolumesUsag
 
         SearchBuilder<VolumeVO> sb =  volumeDao.createSearchBuilder();
         sb.and("idIN", sb.entity().getId(), SearchCriteria.Op.IN);
-        sb.and("name", sb.entity().getName(), SearchCriteria.Op.LIKE);
```
Comment (DaanHoogland) [inline plugins/metrics/src/main/java/org/apache/cloudstack/metrics/MetricsServiceImpl.java:343]: incompatible

File: server/src/main/java/com/cloud/network/vpc/VpcManagerImpl.java [hunk 2]
```diff
@@ -1391,7 +1391,7 @@ public Pair<List<? extends Vpc>, Integer> listVpcs(final Long id, final String v
         final SearchBuilder<VpcVO> sb = vpcDao.createSearchBuilder();
         _accountMgr.buildACLSearchBuilder(sb, domainId, isRecursive, permittedAccounts, listProjectResourcesCriteria);
 
-        sb.and("name", sb.entity().getName(), SearchCriteria.Op.LIKE);
```
Comment (DaanHoogland) [inline server/src/main/java/com/cloud/network/vpc/VpcManagerImpl.java:1394]: incompatible

File: plugins/metrics/src/main/java/org/apache/cloudstack/metrics/MetricsServiceImpl.java [hunk 3]
```diff
@@ -287,7 +287,7 @@ protected Pair<List<VMInstanceVO>, Integer> searchForSystemVmsInternal(ListSyste
 
         SearchBuilder<VMInstanceVO> sb =  vmInstanceDao.createSearchBuilder();
         sb.and("idIN", sb.entity().getId(), SearchCriteria.Op.IN);
-        sb.and("name", sb.entity().getName(), SearchCriteria.Op.LIKE);
```
[the remark this answers is not in the document]
Comment (DaanHoogland) [reply plugins/metrics/src/main/java/org/apache/cloudstack/metrics/MetricsServiceImpl.java:290]: also, but not in scope maybe, searchbuilders do not belong in the service layer. These should be moved to the DAOs

File: server/src/main/java/org/apache/cloudstack/backup/BackupManagerImpl.java [hunk 4]
```diff
@@ -223,7 +223,7 @@ public Pair<List<BackupOffering>, Integer> listBackupOfferings(final ListBackupO
         final Filter searchFilter = new Filter(BackupOfferingVO.class, "id", true, cmd.getStartIndex(), cmd.getPageSizeVal());
         SearchBuilder<BackupOfferingVO> sb = backupOfferingDao.createSearchBuilder();
         sb.and("zone_id", sb.entity().getZoneId(), SearchCriteria.Op.EQ);
-        sb.and("name", sb.entity().getName(), SearchCriteria.Op.LIKE);
```
Comment (DaanHoogland) [inline server/src/main/java/org/apache/cloudstack/backup/BackupManagerImpl.java:226]: incompatible

**Comment by harikrishna-patnala** [comment]
> one problem @harikrishna-patnala , people using a search on name with the API won´t be able to find by partial names anymore

They can use the 'keyword' param @DaanHoogland. Yes I agree already existing scripts/configurations in environments may have issues if they are using APIs with name parameter. But I feel this has to be fixed, for example

1. There are two VMs with names "ABCD" and "ABCDxyz"
2. When I search with name "ABCD" using API both VMs will appear in the list (if name search considers partial names)

If this is the way it has used in some scripts or configuration, it will be difficult to find the VM.

**Comment by harikrishna-patnala** [comment]
> > one problem @harikrishna-patnala , people using a search on name with the API won´t be able to find by partial names anymore
> 
> They can use the 'keyword' param @DaanHoogland. Yes I agree already existing scripts/configurations in environments may have issues if they are using APIs with name parameter. But I feel this has to be fixed, for example
> 
> 1. There are two VMs with names "ABCD" and "ABCDxyz"
> 2. When I search with name "ABCD" using API both VMs will appear in the list (if name search considers partial names)
> 
> If this is the way it has used in some scripts or configuration, it will be difficult to find the VM.

@DaanHoogland any thoughts here

**Comment by DaanHoogland** [comment]
> > > one problem @harikrishna-patnala , people using a search on name with the API won´t be able to find by partial names anymore
> > 
> > 
> > They can use the 'keyword' param @DaanHoogland. Yes I agree already existing scripts/configurations in environments may have issues if they are using APIs with name parameter. But I feel this has to be fixed, for example
> > 
> > 1. There are two VMs with names "ABCD" and "ABCDxyz"
> > 2. When I search with name "ABCD" using API both VMs will appear in the list (if name search considers partial names)
> > 
> > If this is the way it has used in some scripts or configuration, it will be difficult to find the VM.
> 
> @DaanHoogland any thoughts here

yes, too many and contradicting. I was mostly biased against backwards incompatibilities when I first responded. I still tend to, but we might get away with a notice and considder it a fix.

**Comment by shwstppr** [comment]
@blueorangutan package

**Review by shwstppr — APPROVED** [review]
Code LGTM. Makes behaviour consistent with other APIs

**Comment by shwstppr** [comment]
@blueorangutan test

**Review by rohityadavcloud — APPROVED** [review]
LGTM didn't test it, but code looks alright

**Pull request merged**